src/core/Akka/Event/EventBusUnsubscriber.cs

```diff
@@ -45,140 +42,83 @@ public EventStreamUnsubscriber(EventStream eventStream, ActorSystem system, bool
-                Context.Unwatch(unregister.Actor);
-            }).With<Terminated>(terminated =>
-            {
-                if (_debug)
-                    _eventStream.Publish(new Debug(this.GetType().Name, GetType(),
-                        string.Format("unsubscribe {0} from {1}, because it was terminated", terminated.Actor , _eventStream )));
-                _eventStream.Unsubscribe(terminated.Actor);
-            })
-            .WasHandled;
+                case Register register:
+                {
+                    if (_debug)
+                        _eventStream.Publish(new Debug(GetType().Name, GetType(),
+                            $"watching {register.Actor} in order to unsubscribe from EventStream when it terminates"));
+                    Context.Watch(register.Actor);
+                    break;
+                }
+                case UnregisterIfNoMoreSubscribedChannels unregister:
+                {
+                    if (_debug)
+                        _eventStream.Publish(new Debug(GetType().Name, GetType(),
+                            $"unwatching {unregister.Actor} since has no subscriptions"));
+                    Context.Unwatch(unregister.Actor);
+                    break;
+                }
+                case Terminated terminated:
+                {
+                    if (_debug)
+                        _eventStream.Publish(new Debug(GetType().Name, GetType(),
+                            $"unsubscribe {terminated.ActorRef} from {_eventStream}, because it was terminated"));
+                    _eventStream.Unsubscribe(terminated.ActorRef);
+                    break;
+                }
+                default:
+                    return false;
+            }
+
+            return true;
         }
 
-        /// <summary>
-        /// TBD
-        /// </summary>
         protected override void PreStart()
         {
             if (_debug)
-                _eventStream.Publish(new Debug(this.GetType().Name, GetType(),
+                _eventStream.Publish(new Debug(GetType().Name, GetType(),
                     string.Format("registering unsubscriber with {0}", _eventStream)));
-            _eventStream.InitUnsubscriber(Self, _system);
         }
 
         /// <summary>
-        /// TBD
+        /// INTERNAL API
+        ///
+        /// Registers a new subscriber to be death-watched and automatically unsubscribed.
         /// </summary>
         internal class Register
         {
-            /// <summary>
-            /// TBD
-            /// </summary>
-            /// <param name="actor">TBD</param>
             public Register(IActorRef actor)
             {
                 Actor = actor;
             }
 
             /// <summary>
-            /// TBD
-            /// </summary>
-            public IActorRef Actor { get; private set; }
-        }
-
-
-        /// <summary>
-        /// TBD
-        /// </summary>
-        internal class Terminated
```

Comment:
Most probable cause of #5717 - we were handling our own `Terminated` message rather than the built-in one. This looks like a porting error that never was caught years ago.